modules/zarr/src/lib/utils.ts

```diff
@@ -0,0 +1,106 @@
+import { openGroup } from 'zarr';
+import type { ZarrArray } from 'zarr';
+import type { PixelSource } from '../types';
+
+interface Multiscale {
+  datasets: { path: string }[];
+  version?: string;
+}
+
+export interface RootAttrs {
+  multiscales: Multiscale[];
+}
+
+export async function loadMultiscales(store: ZarrArray['store'], path = '') {
+  const grp = await openGroup(store, path);
+  const rootAttrs = (await grp.attrs.asObject()) as RootAttrs;
+
+  let paths = ['0'];
+  if ('multiscales' in rootAttrs) {
+    const { datasets } = rootAttrs.multiscales[0];
+    paths = datasets.map(d => d.path);
+  }
+
+  const data = paths.map(path => grp.getItem(path));
+  return {
+    data: (await Promise.all(data)) as ZarrArray[],
+    rootAttrs
+  };
+}
+
+function prevPowerOf2(x: number) {
+  return 2 ** Math.floor(Math.log2(x));
+}
+
+export function guessTileSize(arr: ZarrArray) {
+  const interleaved = isInterleaved(arr.shape);
+  const [yChunk, xChunk] = arr.chunks.slice(interleaved ? -3 : -2);
+  const size = Math.min(yChunk, xChunk);
+  // deck.gl requirement for power-of-two tile size.
+  return prevPowerOf2(size);
+}
+
+
+/*
+ * The 'indexer' for a Zarr-based source translates
+ * a 'selection' to an array of indices that align to
+ * the labeled dimensions.
+ *
+ * > const labels = ['a', 'b', 'y', 'x'];
+ * > const indexer = getIndexer(labels);
+ * > console.log(indexer({ a: 10, b: 20 }));
+ * > // [10, 20, 0, 0]
+ */
+export function getIndexer<T extends string>(labels: T[]) {
+  const size = labels.length;
+  const dims = getDims(labels);
+  return (sel: { [K in T]: number } | number[]) => {
+    if (Array.isArray(sel)) {
+      return [...sel];
+    }
+    const selection: number[] = Array(size).fill(0);
+    for (const [key, value] of Object.entries(sel)) {
+      selection[dims(key as T)] = value as number;
+    }
+    return selection;
+  };
+}
+
+/*
+ * Helper method to determine whether pixel data is interleaved or not.
+ * > isInterleaved([1, 24, 24]) === false;
+ * > isInterleaved([1, 24, 24, 3]) === true;
```

Comment:
Is the Zarr array shape always in a well-specified order like `x, y, z` (or I'm guessing `x, y, z` doesn't usually have a meaning in Zarr?)? 

Just wondering because with rasterio it's easy to get confused with band order because [it uses](https://rasterio.readthedocs.io/en/latest/topics/image_processing.html#imageorder) 

```
(bands, rows, columns)
```

while image processing software like `scikit-image`, `pillow` and `matplotlib` are generally ordered:

```
(rows, columns, bands)
```

Basically my question is: how do you know that a 3 or 4 in the last dimension means the data is interleaved?

Reply:
A Zarr array is n-dimensional. Zarr.js only handles row-major ordering of arrays ("C" order), so we assume the last 2-3 dimensions are "special". If an image is interleaved, the last three dimension are `(rows, columns, bands)`, otherwise the last two dimensions are assumed to be `(rows, columns)`. The `rows` and `columns` dimensions are down-sampled across resolutions.